spirv_glsl.cpp

```diff
@@ -3221,6 +3239,199 @@ string CompilerGLSL::access_chain(uint32_t base, const uint32_t *indices, uint32
+
+	auto buffer_name = to_name(expression_type(base).self);
+
+	expr += buffer_name;
+	expr += "[";
+
+	uint32_t result = flattened_access_chain_offset(expr, base, indices, count, offset);
+	assert(result % 4 == 0);
+
+	uint32_t index = result / 4;
+
+	expr += convert_to_string(index / 4);
+	expr += "]";
+
+	static const char* swizzle[4][4] =
+	{
+		{ ".x", ".y", ".z", ".w" },
+		{ ".xy", ".yz", ".zw" },
+		{ ".xyz", ".yzw" },
+		{ "" }
+	};
+
+	assert(target_type.vecsize >= 1 && target_type.vecsize <= 4);
+	assert(swizzle[target_type.vecsize - 1][index % 4]);
+
+	expr += swizzle[target_type.vecsize - 1][index % 4];
+}
+
+uint32_t CompilerGLSL::flattened_access_chain_offset(std::string &expr, uint32_t base, const uint32_t *indices, uint32_t count, uint32_t offset)
+{
+	const auto *type = &expression_type(base);
+	uint32_t type_size = 0;
+
+	// For resolving array accesses, etc, keep a local copy for poking.
+	SPIRType temp;
+
+	for (uint32_t i = 0; i < count; i++)
+	{
+		uint32_t index = indices[i];
+
+		// Arrays
+		if (!type->array.empty())
+		{
+			// We have to modify the type, so keep a local copy.
+			if (&temp != type)
+				temp = *type;
+			type = &temp;
+
+			uint32_t array_size = temp.array.back();
+			temp.array.pop_back();
+
+			assert(type_size > 0);
+			assert(type_size % array_size == 0);
+
+			uint32_t array_stride = type_size / array_size;
+			assert(array_stride % 16 == 0);
+
+			expr += to_expression(index);
+			expr += "*";
+			expr += convert_to_string(array_stride / 16);
+			expr += "+";
+
+			type_size = array_stride;
+		}
+		// For structs, the index refers to a constant, which indexes into the members.
+		// We also check if this member is a builtin, since we then replace the entire expression with the builtin one.
+		else if (type->basetype == SPIRType::Struct)
+		{
+			index = get<SPIRConstant>(index).scalar();
+
+			if (index >= type->member_types.size())
+				SPIRV_CROSS_THROW("Member index is out of bounds!");
+
+			offset += type_struct_member_offset(*type, index);
+
+			type_size = uint32_t(get_declared_struct_member_size(*type, index));
+			type = &get<SPIRType>(type->member_types[index]);
+		}
+		// Matrix -> Vector
+		else if (type->columns > 1)
+		{
+			if (ids[index].get_type() == TypeConstant)
+			{
+				auto &c = get<SPIRConstant>(index);
+				offset += c.scalar() * 16;
+			}
+			else
+				SPIRV_CROSS_THROW("Cannot flatten dynamic matrix indexing!");
+
+			// We have to modify the type, so keep a local copy.
+			if (&temp != type)
+				temp = *type;
+			type = &temp;
+			temp.columns = 1;
+		}
+		// Vector -> Scalar
+		else if (type->vecsize > 1)
+		{
+			if (ids[index].get_type() == TypeConstant)
+			{
+				auto &c = get<SPIRConstant>(index);
+				offset += c.scalar() * 4;
+			}
+			else
+				SPIRV_CROSS_THROW("Cannot flatten dynamic vector indexing!");
+
+			// We have to modify the type, so keep a local copy.
+			if (&temp != type)
```

Comment:
Yup - if this was the case it would be possible to basically reconstruct the original SPIRV hierarchy of types which would make this simpler because access chains traverse this hierarchy.